browser.c

```diff
@@ -1606,21 +1606,60 @@ void mutt_buffer_select_file(struct Buffer *file, SelectFileFlags flags,
           mutt_message("%s", ARRAY_GET(&state.entry, menu->current)->name);
         break;
 
-#ifdef USE_IMAP
-      case OP_BROWSER_TOGGLE_LSUB:
-        bool_str_toggle(NeoMutt->sub, "imap_list_subscribed", NULL);
+      case OP_CREATE_MAILBOX:
+      {
+        int create_rc;
+        bool created = false;
+        struct Mailbox *m = mailbox_find(CurrentFolder);
+        /* PATH_MAX used for local mailboxes, e.g. maildir, but should also be
+         * sufficient for remote mailboxes */
+        char path[PATH_MAX] = { '\0' };
 
-        mutt_unget_event(0, OP_CHECK_NEW);
-        break;
+        if (state.imap_browse)
+        {
+#ifdef USE_IMAP /* leaving this until mbox_create implemented for IMAP */
+          create_rc = imap_mailbox_create(mutt_buffer_string(&LastDir));
+#endif /* USE_IMAP */
+        }
+        else if (m)
+        {
+          if (m->mx_ops->mbox_create == NULL)
+          {
+            mutt_error(_("Create is not supported for %s mailboxes"), m->mx_ops->name);
+            break;
+          }
 
-      case OP_CREATE_MAILBOX:
-        if (!state.imap_browse)
+          /* No particularly clear that this is created under the current mailbox */
+          mutt_get_field(_("Create mailbox: "), path, sizeof(path),
+                         MUTT_COMP_NO_FLAGS, false, NULL, NULL);
+          struct Mailbox *new_mailbox = mutt_mem_malloc(sizeof(struct Mailbox));
+
+          create_rc = m->mx_ops->mbox_create(m->account, m, path, &new_mailbox);
+        }
+
+        switch (create_rc)
         {
-          mutt_error(_("Create is only supported for IMAP mailboxes"));
-          break;
+          case MX_CREATE_OK: /* = 0 (for imap_mailbox_create) */
+            mutt_message(_("Mailbox created"));
+            created = true;
+            break;
+          case MX_CREATE_EXISTS:
+            mutt_error(_("A %s mailbox already exists under %s"), m->mx_ops->name, path);
+            break;
+          case MX_CREATE_BAD_NAME:
+            mutt_error(_("Bad name '%s' (is it empty?)"), path);
+            break;
+          case MX_CREATE_SYS_ERROR:
+            mutt_error(_("Mailbox creation failed: %s"), strerror(errno));
+            break;
+          case MX_CREATE_ERROR:
+          case -1: /* imap_mailbox_create */
+            mutt_error(_("Mailbox creation failed"));
+            break;
         }
 
-        if (imap_mailbox_create(mutt_buffer_string(&LastDir)) == 0)
+#ifdef USE_IMAP
+        if (created)
         {
           /* TODO: find a way to detect if the new folder would appear in
            *   this window, and insert it without starting over. */
```

Comment:
Not sure how much (currently none) of this I should be using. There is some coupling to IMAP, but I think that could be filtered?

Reply:
Hmm...  I think we'll need most of it, to start with.

When create/delete/rename Mailbox succeeds, we need to refresh our view.
This code does it by completely tearing down the view and rebuilding it.

The architecturally-cleaner solution, which we can get to later, is:
- Browser sets up observers for the objects it's displaying
- User performs a function
- Backend performs the actual work
- Backend sends a notification of change
  e.g. Account: new Mailbox
- Browser receives notification and updates itself (non-destructively :-)

Reply:
> Hmm... I think we'll need most of it, to start with.

Yeah, I think I'll need to dive into `imap_browse` .... it seems to be doing most of the setup here